// @sudoku/gameStore.js
/*
  Game Store Adapter - 将领域对象接入 Svelte 响应式系统
  
  职责：
  - 持有 Game / Sudoku 领域对象
  - 对外暴露 Svelte writable store 响应式状态
  - 对外暴露 UI 可调用的方法（guess, undo, redo, hint, explore 等）
  - 当领域对象变化时，自动更新 store 触发 UI 刷新
  
  设计说明：
  - 使用 Svelte writable store 作为桥接层
  - 领域对象变化时通过 subscribe 机制触发 store.update()
  - UI 组件通过 $store 语法消费响应式状态
  - gameStore.grid 成为 UI grid 数据的唯一来源
  - 添加 hint 和 explore 相关响应式状态
 */

import { writable } from 'svelte/store';
import { createGame, createSudoku, createGameFromJSON } from './domain/index.js';

export function createGameStore() {
  // 内部持有 Game 领域对象
  let game = null;
  // 保存取消订阅函数，防止内存泄漏
  let unsubscribeGame = null;

  // 响应式状态
  const grid = writable(Array(9).fill(null).map(() => Array(9).fill(0)));
  const invalidCells = writable([]);
  const canUndo = writable(false);
  const canRedo = writable(false);
  const isComplete = writable(false);
  const isExploring = writable(false);
  const candidates = writable([]); // 当前选中单元格的候选数
  const hint = writable(null); // 当前提示

  // 从领域对象同步状态到 store
  function syncToStores() {
    if (!game) return;

    const sudoku = game.getSudoku();
    const currentGrid = sudoku.getGrid();
    grid.set(currentGrid);

    // 计算冲突单元格
    const invalid = sudoku.validate();
    invalidCells.set(invalid);

    canUndo.set(game.canUndo());
    canRedo.set(game.canRedo());
    isComplete.set(sudoku.isComplete());
    isExploring.set(game.isInExploration());
  }

  // 取消当前游戏的订阅
  function unsubscribeCurrent() {
    if (unsubscribeGame) {
      unsubscribeGame();
      unsubscribeGame = null;
    }
  }

  return {
    // 响应式状态（可被 UI 订阅）
    grid,
    invalidCells,
    canUndo,
    canRedo,
    isComplete,
    isExploring,
    candidates,
    hint,

    // 初始化新游戏
    initGame(initialGrid) {
      // 取消旧游戏的订阅
      unsubscribeCurrent();

      const sudoku = createSudoku(initialGrid);
      game = createGame({ sudoku });

      // 订阅领域对象的变化，并保存取消订阅函数
      unsubscribeGame = game.subscribe(() => {
        syncToStores();
      });

      syncToStores();
    },

    // 获取当前 Game 对象（用于序列化等）
    getGame() {
      return game;
    },

    // 从 JSON 恢复游戏
    loadFromJSON(json) {
      // 取消旧游戏的订阅
      unsubscribeCurrent();

      game = createGameFromJSON(json);

      unsubscribeGame = game.subscribe(() => {
        syncToStores();
      });

      syncToStores();
    },

    // 用户输入
    guess(row, col, value) {
      if (!game) return;
      game.guess({ row, col, value });
    },

    // 撤销
    undo() {
      if (!game) return;
      game.undo();
    },

    // 重做
    redo() {
      if (!game) return;
      game.redo();
    },

    // HW2: 获取候选数
    getCandidates(row, col) {
      if (!game) return [];
      return game.getCandidates(row, col);
    },

    // HW2: 获取下一步提示
    getNextHint() {
      if (!game) return null;
      return game.getNextHint();
    },

    // HW2: 使用提示
    useHint() {
      if (!game) return;
      const nextHint = game.getNextHint();
      if (nextHint) {
        game.guess(nextHint);
      }
    },

    // HW2: 进入探索模式
    enterExploration() {
      if (!game) return;
      game.enterExploration();
    },

    // HW2: 退出探索模式（不保存更改）
    exitExploration() {
      if (!game) return;
      game.exitExploration();
    },

    // HW2: 提交探索结果（保存更改）
    submitExploration() {
      if (!game) return;
      game.submitExploration();
    },

    // HW2: 检查是否冲突
    isConflicted() {
      if (!game) return false;
      return game.isConflicted();
    },

    // 检查是否可以撤销
    canUndo() {
      return game ? game.canUndo() : false;
    },

    // 检查是否可以重做
    canRedo() {
      return game ? game.canRedo() : false;
    },

    // 检查是否在探索模式中
    isInExploration() {
      return game ? game.isInExploration() : false;
    }
  };
}

// 创建单例 game store
export const gameStore = createGameStore();
